### PR TITLE
Highlight row on mouseover

### PR DESCRIPTION
**Pull request by kawashita86**
Hello,
first of all thanks for developing this app, it's amazing and each upgrade is for the better!
As i found that highlighting a row was a requested feature (https://github.com/casperstorm/ajour/issues/241) I've tried to implement it. Since i'm a newbie in Rust, and just started diving into the language, I went for an easy solution that can be helpful for the eye.

Resolves #

## Proposed Changes
  -  Implemented a custom widget "TableRow" that is a superset of iced_native::container and that allows a "hovered" state.
  -  Replaced the type of element::catalog_data_cell and element::addon_data_cell from Container to TableRow
  - Added custom styling for the new type by following the guideline of the other element in gui/style.rs to show a border around  hovered row 
image

## Checklist

- [x] Tested on Windows
- [x] Tested on MacOS
- [ ] Tested on Linux
- [ ] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users

### CODE REVIEW

**Comment by tarkah** [comment]
```
> new to rust
> wanted easy feature
> adds custom widget
```

Wow! This is really awesome! Great stuff @kawashita86. You definitely didn't choose the easiest thing to implement :P

Can you run `rustfmt` and fix the lints with `clippy``? I'll dig into this more later today, but at first glance looks like the right approach.

**Comment by tarkah** [comment]
Also, can you rebase to development branch? We upgraded our iced dependencies to 0.2 release and there are also some small api changes inside widget due to that.

**Comment by kawashita86** [comment]
Sure, let me rebase and push it

**Comment by casperstorm** [comment]
First of all, great first contribution. Wow.
I just tried it out and it will be a nice addition, very requested.

Heres a few things from a UI perspective:

- We have a left and right padding around content, I think we should wrap your highlightable row into a new row maybe so we don't highlight this padding.
- Colors is almost there, right now I think it is a little too distracting. I would like you to maybe try using the faded color for border as well, so its the same as your highlighted background. This is of cause just me suggesting.

**Comment by kawashita86** [comment]
@casperstorm Thanks again for the kind words, glad that the addition is useful.
I completely agree with your feedback and I will try to get it done with some more coding, i will push them to this branch as soon as i've got them fixed.

File: crates/widgets/src/widget/table_row.rs
Comment (tarkah) [inline crates/widgets/src/widget/table_row.rs]: You should return the status from `self.content.on_event` since it needs to bubble up

**Review by tarkah — CHANGES_REQUESTED** [review]

File: crates/widgets/src/widget/table_row.rs
Comment (kawashita86) [reply crates/widgets/src/widget/table_row.rs]: Thanks for the suggestion in the next commit i'm fixing it.

**Comment by kawashita86** [comment]
Hello,
with the last commit i should have improved a bit the layout of the table_row widget:
- reduced the hovered contrast by changing the opacity of the border to a softer value
- added new property to the style of the table_row that define the offset of the row within the bounds of the layout, so now the hovered state should not get out of bounds and can be controlled.
- now the "on_event" function return the self.content.on_event status as suggested by @tarkah

**Comment by casperstorm** [comment]
> with the last commit i should have improved a bit the layout of the table_row widget:

Awesome job once again. It looks nice. Here is a side-by-side from development and your branch:

highlight

- Would you try to work on getting us back to 1px between rows? I think it looks cleaner but it might not be easy so lets see what options we have to get there.
- I still feel we can optimize the coloring a bit. Optimal I would like us to keep the addon highlighting color similar to when you hover a column title: border color and background is the same. I feel it will create a better harmony.
- When you see a highlightable row/cell on our smart phone we are used to being able to click it. This is not the case right now. We have to click the title in order to expand it. Going into the future I would like us to consider if its possible that the whole row is clickable, while still being able to add buttons like "Update" which wouldn't trigger the cell click handler.

Again small points of feedback you can pick up if you have the time. Again thank you for doing this and welcome to the team :)

**Comment by kawashita86** [comment]
Based on the suggestion i've decided to rework a bit the idea behind a current table row:
the are now two new style implementation  "HoverableBrightForegroundContainer" and "HoverableForegroundContainer" these are similar to the old "BrightForegroundContainer" and "NormalForegroundContainer" but they don't have a background set, so now the table_row is in charge of displaying the correct background color for the entire row, both in hovered and normal state. 
This should also make it easier to create further styling by having a single entrypoint, and should adhere more to the original layout design.

> When you see a highlightable row/cell on our smart phone we are used to being able to click it. This is not the case right now. We have to click the title in order to expand it. Going into the future I would like us to consider if its possible that the whole row is clickable, while still being able to add buttons like "Update" which wouldn't trigger the cell click handler.

I totally agree with your idea, I will study a bit this and see if there is a clean approach to solve the event propagation and capture as I fear that it can get easily messy without a good implementation.

**Comment by tarkah** [comment]
@kawashita86 you can check the return of `self.content.on_event` and if it is `Status::Ignored`, then you check query the event to see if it's a left click. If so, you can append an `on_click` message into `messages`. You should add on `on_click` builder method similar to the `on_resize` one I have for the header widget. I hope that makes sense.

**Comment by kawashita86** [comment]
@tarkah Thanks for pointing out a way to achieve the wanted behaviour.
I was working on something really similar, but i was slowed down a bit since when the row is expanded (on MyAddon tab) the pickup list for the release channel is propagating back the `event::Status::Ignored` creating a weird behaviour of the row opening and closing any time you interact with the picklist.
I've decided to use  a `custom_bounds` Rectangle inside the `on_event` that is limited to the bounds of the closed row instead of the expanded one.
I will wait for your feedback to move on.

File: crates/widgets/src/widget/table_row.rs
Comment (tarkah) [inline crates/widgets/src/widget/table_row.rs:79]: I'd reword this to explain this is the max height of the "clickable" area

**Review by tarkah — CHANGES_REQUESTED** [review]
This looks really good! I'd change the the hovered style to only trigger based on the custom bounds as well. That way the expanded row doesn't show as highlighted when interacting with it. Then the pick-list issue isn't even a problem. If we want the expanded row to have different coloring, we should handle that styling based on the fact it's expanded, not based on the "hover" of this row.

**Comment by tarkah** [comment]
I'm running into a couple weird bugs w/ this PR. The first, there is a visual artifact at the top between the flavor selector and version:

image

The second.. if I left click on the scrollbar, it gets "stuck" and follows my mouse movement around. What I expect is happening is the scrollbar captures the left click, but doesn't capture the release, because our row is taking precedence maybe? Similar to how you have the event capturing only taking place after a certain height, can you lock the event capturing to be only inside the bounds of the actual rectangle?

File: crates/widgets/src/renderer/table_row.rs
Comment (tarkah) [inline crates/widgets/src/renderer/table_row.rs:39]: We shouldn't hardcode this offset to the implementation of this widget. What if we want to use this widget, but have it highlight the entire thing? We should just wrap this widget in a Row widget and apply the padding there by pushing some Space on each side of it.

**Review by tarkah — CHANGES_REQUESTED** [review]

File: crates/widgets/src/renderer/table_row.rs
Comment (tarkah) [reply crates/widgets/src/renderer/table_row.rs:39]: Ahh, I see we can specify it with the style, so I guess not that big of a deal. I'm ok leaving it this way.

Comment (tarkah) [reply crates/widgets/src/renderer/table_row.rs:39]: And when I think about it more, we would need this anyway to work well with the nested scrollable since we don't want the scrollbar to get highlighted when hovering over it, so a offset for the styling would be needed in that case

Comment (tarkah) [reply crates/widgets/src/renderer/table_row.rs:39]: Omg I'm retarded, disregard everything I said. This is inside the scollable, not the other way around :P

**Comment by tarkah** [comment]
Ok, here is a fresh comment since I was all over the place :P So the Scrollable widget is bugged, in that it resolves nested events _before_ the events on the scrollbar, even though the scrollbar should take precedence over any widget nested inside the scrollable since it's technically the top layer of the nested contents. I'll open a PR w/ Iced so they can fix this behavior.

The fix for us is we need to not return `Status::Captured` on left mouse button release of our TableRow widget as it prevents the scrollbar from resolving on that event, which thing locks the scrollbar as captured.

**Comment by tarkah** [comment]
Opened issue here: https://github.com/hecrj/iced/issues/649

**Comment by kawashita86** [comment]
@tarkah first of all thank you for taking time to go through everything in details, as you guessed I've already implemented a public method to set the height of the row used for the custom bound, probably `inner_row_height` is not the most fortunate choice ever.
I'm not happy with the idea of having to set manually the height of a property that is used mostly for internal behaviour, but it was the best idea i could find to solve the pick-list issue without creating a monster.
So as you suggest i will change the current render method to apply the hovered style to the custom_bounds only and change the event::Status::Captured to circumvent the current iced bug.

**Comment by tarkah** [comment]
Great! An for the `menu_container` can you please revert the style for those back to `NormalForegroundContainer`? I believe that's what's causing the visual bug in the top menu area. I assume this was overridden with a find / replace?

**Comment by kawashita86** [comment]
Ok so i've committed the latest changes, i should also have reverted the old `menu_container` style to the previous `BrightForegroundContainer`.

File: src/gui/element.rs
Comment (tarkah) [inline src/gui/element.rs:1726]: Still one that needs to be changed here. Is everything else good to go outside this change? If so, I'll do a final test before merging.

**Review by tarkah — CHANGES_REQUESTED** [review]

**Review by kawashita86 — COMMENTED** [review]
I've reviewed all the request opened

**Review by tarkah — APPROVED** [review]
Great changes. This looks really nice! @casperstorm all looks good to me

**Review by casperstorm — CHANGES_REQUESTED** [review]
One last thing I have is to convert addon title from a Button to a Text. We don't need to have that now that we can (FINALLY!) press the row it self to expand. Should be super easy, basically just convert this to a `Text::new` (or see how its done for `installed_version`). The relevant code is right here:

https://github.com/casperstorm/ajour/blob/90feb56bc31b0d6d1fa978513f41957d0e3a402e/src/gui/element.rs#L739-L740

@kawashita86 you have done amazingly. I can't believe how great this PR has turned out to be. My testing found no issues.

**Comment by kawashita86** [comment]
> One last thing I have is to convert addon title from a Button to a Text. We don't need to have that now that we can (FINALLY!) press the row it self to expand. Should be super easy, basically just convert this to a `Text::new` (or see how its done for `installed_version`). The relevant code is right here:
> 
> https://github.com/casperstorm/ajour/blob/90feb56bc31b0d6d1fa978513f41957d0e3a402e/src/gui/element.rs#L739-L740
> 

Ok, I've reworked a bit how a row is built as you suggested:
- Removed the button from both the "My Addons" tab and the "Catalog", now they are simple `Text` elements.
- Removed the ` website_state: button::State` from `CatalogRow` as it was not used anymore.
- Added in the `table_row` renderer the` mouse::Interaction::Pointer` when the mouse is in hovered state, to make it easier to understand that there's a possible action by clicking the row.

It should be everything now.

**Review by casperstorm — APPROVED** [review]

**Comment by casperstorm** [comment]
@kawashita86 Amazing, just tested and everything works nicely.
Do you want to update CHANGELOG.md or should I?

**Comment by kawashita86** [comment]
@casperstorm Glad that everything works fine!! Please go on with the Changelog entry, i'm SO bad at writing commit messages and changelog  that i'm always happy when someone else takes the task.

**Pull request merged**
